llext: support section alignment requirements

pillo79 opened the pull request:
This PR adds support in LLEXT for symbol / section alignment requirements.

Symbols that require precise alignment will cause the compiler to generate proper padding inside each section so that they are properly placed; also, the section itself will be flagged to be aligned to the largest alignment requirement of any symbol in the section. The ELF spec also guarantees that symbol alignment is respected in the file offsets, so a suitably placed ELF file could be directly used.

* When LLEXT maps similar sections into a contiguous region, each section's alignment is checked to ensure offsets and alignments are coherent, so that every section will be properly allocated once the region is fully defined. 
 Special care is taken when a section is appended to a region which starts at an address that does not match the current aligment requirements. In that case, the region start must be moved backwards to satisfy the new, stricter alignment requirements, and this may result in false positives in the overlap detection.    

* Finally, memory is allocated also taking into account the existing MMU/MPU requirements, if applicable. When the ELF file buffer can be directly reused, the section addresses are simply checked for compliance.

### Review notes

* The first 4 commits are preparatory search-and-replace that introduce no functional changes (except for some log messages). Might be useful to view those ignoring whitespace due to large indentation reductions. 
* Commits 5, 6 and 7 are refactors of existing code that minimize the changes in the following commits.
* The new features for this PR are implemented in the last two commits.

pillo79 committed:
llext: (cosmetic) cleaner region overlap checks

Rewrite the overlap checks in llext_map_sections() to use a pair of
macros to calculate the top and bottom of a region. This makes the
code more readable and less error-prone.

No functional change is introduced in this commit apart from a log
message format change.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
**v2**:
 * fixed a bug in the overlap checks refactor that made the test ignore partial overlaps - previous behavior restored
 * applied suggestion from @lyakh and used a direct calculation for the required section size/alignment in the `CONFIG_ARM_MPU` case 
 * reworded a message that was suggesting a workaround was always possible (... it's not in the SOF `pre_located` case)

pillo79 commented:
DNM since an issue was identified during SOF tests and a separate fix needs to be pushed before this could be merged.

teburd commented:
If SoF is using llext in some manner that isn't tested by Zephyr's testing we are hopelessly going to be falling into the trap of breaking SoF.

teburd commented on subsys/llext/llext_load.c:
```diff
@@ -402,16 +402,18 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
-						"overlaps with %d (0x%zx +%zd)",
-						i, (size_t)x->sh_addr, (size_t)x->sh_size,
-						j, (size_t)y->sh_addr, (size_t)y->sh_size);
+#define BOT(reg) reg->sh_addr
```
At this point would it make more sense to check for overlap in a static inline with a self documenting function name?

static inline bool regions_overlap(struct ... reg0, struct ...reg1);

if (regions_overlap(x, y)) { LOG_ERR("Region ...  overlaps ..." ... ) }

pillo79 replied:
They are actually working on different fields (one on `sh_addr`, one on `sh_offset`) so it's more like a macro. But yeah I'll try to find a clearer way for that check! :-)

teburd commented on subsys/llext/llext_load.c:
```diff
@@ -423,16 +425,18 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
-					i, (size_t)x->sh_offset, (size_t)x->sh_size,
-					j, (size_t)y->sh_offset, (size_t)y->sh_size);
+#define BOT(reg) reg->sh_offset
+#define TOP(reg) (reg->sh_offset + reg->sh_size - 1)
```
a regions_overlap function would make this cleaner in both places as noted before

teburd reviewed (commented):
Some minor comments around macro usage here, but the changes to me look great. I'm assuming this has been tested on NXP hardware as well where the MPU isn't quite like the ARM MPU? E.g. kinetis parts. I found issues here previously that should be verified.

pillo79 committed:
llext: add test for section and symbol alignment

Add a test case for the alignment support in loadable extensions. This
test case creates a set of constants with specific alignment requirements
and verifies that they are placed in memory as expected.

Fix the detached section test to use a more standard syntax for the
section attribute, avoiding issues with different toolchains.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
SOF has now been converted to access extension layout details via the inspection API in https://github.com/thesofproject/sof/pull/9831. This new push adjusts those APIs to compensate for the alignment changes, and so should not have any adverse effect on it. I have ran the SOF CI and it _looks_ okay to me, but I'm holding my breath for a more competent analysis :wink:

Will wait for other pending LLEXT PRs to be merged before rebasing this a final time and re-running both CIs anyway.

pillo79 committed:
llext: avoid redundant 'ldr_parm' checks

Since commit 0aa6b1c9defe, the 'ldr_parm' pointer is guaranteed to be
valid inside all the functions of the llext_load() call tree.

This commit fixes the only exception of llext_copy_strings(), which was
not passed the 'ldr_parm' pointer, and remove the redundant checks.

No functional change is intended by this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

lyakh commented on subsys/llext/llext_mem.c:
```diff
@@ -76,15 +76,23 @@ static int llext_copy_region(struct llext_loader *ldr, struct llext *ext,
-				ext->mem_on_heap[mem_idx] = false;
-				return 0;
+				if (!IS_ALIGNED(ext->mem[mem_idx], region_align) &&
+				    !(ldr_parm && ldr_parm->pre_located)) {
```
you could revert the condition and remove the `else` because now the `if` would contain a `return`

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -398,6 +405,7 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
 					address -= prepad;
 				}
 				bot_ofs -= prepad;
+				region->sh_info = prepad;
```
maybe a single line comment here too, referring to the detailed explanation above

teburd commented on subsys/llext/llext_mem.c:
```diff
@@ -93,41 +96,43 @@ static int llext_copy_section(struct llext_loader *ldr, struct llext *ext,
+
+		block_size = 1 << LOG2CEIL(block_size); /* align to next power of two */
+		region_alloc = block_size;
+		region_align = block_size;
```
just to be clear, there's an ifdef here to deal with page vs mpu region alignment

lyakh replied:
> This is the way ARM's MPU works, the MPU regions need to be aligned based on the region size. https://developer.arm.com/documentation/dui0552/a/cortex-m3-peripherals/optional-memory-protection-unit/mpu-region-base-address-register?lang=en

thanks, didn't know that

pillo79 committed:
llext: (cosmetic) llext_map_sections() cleanups

Add a 'continue' statement when the first section of a region is
detected, to avoid unnecessary indentation.

No functional change is introduced in this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: support alignment requirements for sections

This patch adds support for alignment requirements for sections in the
loader. The alignment requirements are taken from the ELF file and
checked when the section is mapped to a memory region and when the
region is copied to memory (or directly used from the ELF buffer).

Supporting a larger alignment requirement than the region's current
alignment is done by adjusting the region's start address. This needs
care to avoid detecting false overlaps with other regions and to ensure
that the region address and size reported by the inspection APIs are
correct. The extra offset needed to re-adjust these values is stored in
the 'sh_info' field of the region descriptor.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Update still before final rebase to apply review comments. I also cleaned up lots of `ldr_parm != NULL` checks, which were now redundant everywhere (but one sneaky exception!) since https://github.com/pillo79/zephyr/commit/0aa6b1c9defeba7b7622189e43ecaeae9902819e was applied.

pillo79 committed:
llext: (cosmetic) log uniformity cleanups

Use the more compact %#x format for printing out hexadecimal values in
the LLEXT subsystem. This also uses a clearer wording for the
llext_link() log message that avoids repeating information.

No functional change is introduced by this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Rebased after other LLEXT PRs have been merged, [SOF CI test passed](https://sof-ci.01.org/sofpr/PR9860/build11406/devicetest/index.html), ready for final review.

pillo79 committed:
llext: avoid direct llext_string() usage

Since 3466dab8047a the generic llext_symbol_name() function abstracts
the use of llext_string() for (non-section) symbols. Define a similar
llext_section_name() function and replace current occurrences of
llext_string() with the proper abstraction.

By extending llext_symbol_name(), this commit also allows to print the
correct name of sections that are referred to by a symbol.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: (cosmetic) rename llext_copy_section() to llext_copy_region()

llext_copy_section() is a misnomer, as it is copying a region of memory,
not a section. Rename it to llext_copy_region() and use 'region' in the
internal variables to better reflect their purpose.

No functional change is introduced in this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: refactor llext_copy_region()

This commit adjusts the logic of the LLEXT memory allocation routine by:

 * defining 'align' and 'alloc' variables once and updating them in the
   following code;

 * swapping the condition, converting it to an 'if(IS_ENABLED(...))' and
   using a direct calculation for the size to make the code easier to
   read;

 * updating and extending a few comments.

No functional change is intended.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Apologies, but I realized my rebase removed from `llext_link()` a call to a very useful addition by 3466dab8047a,  the new `llext_symbol_name()` wrapper:
https://github.com/zephyrproject-rtos/zephyr/blob/3466dab8047ac4d90630ed3afeb4309ef0ca8c60/include/zephyr/llext/llext_internal.h#L46-L50
so I fully committed to that approach:
* added a similar `llext_section_name()` that applies to section headers;
* applied this PR's change to obtain section names to `llext_symbol_name()`;
* replaced `llext_string()` calls to use the appropriate symbol/section wrapper.

This way all LLEXT code automatically inherits proper object names.

lyakh reviewed (commented):
looks like "llext: support alignment requirements for sections" introduces a potential breakage, which is then fixed by the next commit, but this breaks bisection. Would be better to split differently to avoid the intermittent breakage

pillo79 commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -43,11 +43,24 @@ static inline uintptr_t llext_get_reloc_instruction_location(struct llext_loader
+	return llext_string(ldr, ext, LLEXT_MEM_SHSTRTAB, shdr->sh_name);
+}
+
+static inline const char *llext_symbol_name(const struct llext_loader *ldr,
```
No, because they use different structures (`elf_sym_t` vs `elf_shdr_t`). There are places in the LLEXT code where we loop on section headers, and there's no associated symbol, so both are useful.

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -370,6 +371,38 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
+			 * the original constraint. Rounding the region start
+			 * address 4 bytes down will fix the issue.
+			 */
+			size_t prepad = bot_ofs - ROUND_DOWN(bot_ofs, shdr->sh_addralign);
```
It does indeed appear so, doesn't it? :man_facepalming:
Will fix in the next iteration.

pillo79 commented on subsys/llext/llext_load.c:
```diff
@@ -370,6 +371,38 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
+
+				address -= prepad;
+			}
+			bot_ofs -= prepad;
```
Yes and no. If you compared each region's `sh_offset`/`sh_addr` and `sh_size` fields you would potentially see overlaps, but the used adjustment value is stored in `sh_info` exactly for that purpose. 
The values exported to the user, and those used in region overlap logic, are always correct. Also, prepad areas are zeroed out when loaded to avoid data creep.

pillo79 reviewed (commented):
> looks like "llext: support alignment requirements for sections" introduces a potential breakage, which is then fixed by the next commit, but this breaks bisection. Would be better to split differently to avoid the intermittent breakage.

Indeed that commit fails CI without the next one (RISC-V builds cause section overlap when alignment is considered but not accounted for). However, I feel that distinction is really useful to understand the logic behind the code.
Will go back at the drawing board to try and cut things differently so that every commit is valid, without squashing too much together.

pillo79 commented:
> [...] Would be better to split differently to avoid the intermittent breakage.

Since the code updates the regions array iteratively, the issue appears as soon as I add the alignment code in the loop. To allow bisectability I merged the two commits together and added more comments in the commit message and code.

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -372,6 +381,55 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
+			 * be repeated:
+			 *
+			 *   current region: ofs  0x20, size 0x100, align  0x10
+			 *      new section: ofs 0x200, size  0x20, align 0x100
```
Thanks for adding the layout after the adjustment, this is very helpful! Now I can clearly see that there's something that I don't understand: why should the current region's size be extended to 0x100? We moved the beginning down to 0x20, so we add 4 bytes to the size, which is now 0xe0. And why should the new section now begin at 0x200? Why not at 0x100? What I would expect here is that the new region alignment is set to 0x100 and the `.sh_info` records the fact that the region has to start at 0x24 bytes into the allocated buffer. So, you then allocate a 0x100 byte aligned buffer and copy the region at the 0x24 byte offset into it.

lyakh commented on subsys/llext/llext_load.c:
```diff
@@ -372,6 +381,55 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
+			 */
+			size_t prepad = bot_ofs & (shdr->sh_addralign - 1);
+
+			region->sh_info += prepad;
```
I also don't quite understand this. In my understanding
1. we use the fact that ELF files group sections by their flags
2. so we expect to find a contiguous sequence of compatible sections that we group into a region
3. we also know (or assume) that within the file all sections are aligned
4. we copy regions in one go
5. let's say, using the above example, that this sequence of sections, that we identify as a single region, starts at 0x24524 and the largest alignment in it is 0x100. So, I would expect the result of the alignment-preservation algorithm to be, that we need a 0x100-byte aligned buffer and that we should copy the region into it at the (0x24524 & (0x100 - 1)) = 0x24 byte offset. 

That's it. Just find the largest alignment, take the part of the starting offset masked with that alignment -1, and copy the region there? I don't understand why you add padding here?

lyakh reviewed (changes requested):
I might be missing something... But let me pause this a bit until my doubts are resolved

lyakh reviewed (approved):
I know, that my opinion doesn't always reflect the majority opinion, but looks much simpler and clearer to me, thanks!

teburd commented:
> I know, that my opinion doesn't always reflect the majority opinion, but looks much simpler and clearer to me, thanks!

Your opinion and reviews greatly matter as SOF is a public user of this whole thing. Don't discount this.
